QRCoder/QRCodeGenerator.cs

```diff
@@ -910,72 +989,81 @@ private static bool IsInRange(char c, char min, char max)
             for (var i = 1; i <= numEccWords - 1; i++)
             {
-                var multiplierPolynom = new Polynom();
-                multiplierPolynom.PolyItems.AddRange(new[]{
-                   new PolynomItem(0,1),
-                new PolynomItem(i,0)
-                });
+                multiplierPolynom.PolyItems.Clear();
+                multiplierPolynom.PolyItems.Add(new PolynomItem(0, 1));
+                multiplierPolynom.PolyItems.Add(new PolynomItem(i, 0));
 
                 generatorPolynom = MultiplyAlphaPolynoms(generatorPolynom, multiplierPolynom);
             }
 
             return generatorPolynom;
         }
 
-        private static List<string> BinaryStringToBitBlockList(string bitString)
+        private static byte[] BinaryStringToBitBlockByteList(BitArray bitString, int offset, int count)
         {
             const int blockSize = 8;
-            var numberOfBlocks = (int)Math.Ceiling(bitString.Length / (double)blockSize);
-            var blocklist = new List<string>(numberOfBlocks);
-
-            for (int i = 0; i < bitString.Length; i += blockSize)
+            if (count % blockSize != 0)
+                throw new ArgumentException("Count must be a multiple of 8.", nameof(count));
+            var numberOfBlocks = count / blockSize;
+            var blocklist = new byte[numberOfBlocks];
+
+            int j = 0;
+            count += offset;
+            for (int i = offset; i < count; i += blockSize)
             {
-                blocklist.Add(bitString.Substring(i, blockSize));
+                blocklist[j++] = (byte)(
+                    (bitString[i] ? 128 : 0) + 
+                    (bitString[i + 1] ? 64 : 0) + 
+                    (bitString[i + 2] ? 32 : 0) + 
+                    (bitString[i + 3] ? 16 : 0) + 
+                    (bitString[i + 4] ? 8 : 0) + 
+                    (bitString[i + 5] ? 4 : 0) + 
+                    (bitString[i + 6] ? 2 : 0) + 
+                    (bitString[i + 7] ? 1 : 0));
             }
 
             return blocklist;
         }
 
-        private static List<int> BinaryStringListToDecList(List<string> binaryStringList)
-        {
-            return binaryStringList.Select(binaryString => BinToDec(binaryString)).ToList();
-        }
-
-        private static int BinToDec(string binStr)
-        {
-            return Convert.ToInt32(binStr, 2);
-        }
-
-        private static string DecToBin(int decNum)
+        private static int BinToDec(BitArray bitArray, int offset, int count)
         {
-            return Convert.ToString(decNum, 2);
+            var ret = 0;
+            for (int i = 0; i < count; i++)
+            {
+                ret ^= bitArray[offset + i] ? 1 << (count - i - 1) : 0;
+            }
+            return ret;
         }
 
-        private static string DecToBin(int decNum, int padLeftUpTo)
+        private static void DecToBin(int decNum, int bits, BitArray bitList, ref int index)
```

Comment:
Can this become
```suggestion
        private static void DecToBin(int decNum, int bits, BitArray bitList, int index)
        {
            // loop the same
            return index;
        }
```
Because of the `ref int index` on each increment the memory (stack) must be updated, otherwise it can be in a register.
```diff
- inc dword ptr [r9]
+ inc r9d
```

Reply:
Currently the functionality relies on the caller's `index` variable being added.  We could move it out; then the caller would have to do this:

```cs
DecToBin(value, 8, bitArray, index);
index += 8;
DecToBin(value2, 8, bitArray, index);
index += 8;
```

versus

```cs
DecToBin(value, 8, bitArray, index);
DecToBin(value2, 8, bitArray, index);
```

Reply:
```c#
index = DecToBin(value, 8, bitArray, index);
```
`index` is passed in, worked on it, then returned with the proper value. That's best for perf.

If you want to keep the `ref int index` then use such an approach:
```c#
void Foo(ref int index)
{
    int idx = index;
    ...
    // use idx here, not index
    ...
    index = idx;
}
```
That way `idx` can be a register, and at the end of the method the value of `idx` is written back to the memory `index`.
But perf isn't as good, as `index` must be memory accessible, due to the `ref`, so can't be in a cpu-register.

Reply:
Done.  Nicer to use too.